com.microsoft.mrtk.input/Visualizers/ControllerVisualizer/ControllerModelLoader.cs

```diff
@@ -0,0 +1,87 @@
+using System.Collections.Generic;
+using UnityEngine;
+using System.Threading.Tasks;
+
+
+#if MROPENXR_PRESENT && (UNITY_STANDALONE_WIN || UNITY_WSA || UNITY_ANDROID) && GLTFAST_PRESENT && KTX_PRESENT
+using Microsoft.MixedReality.OpenXR;
+using GLTFast;
+#endif // MROPENXR_PRESENT && (UNITY_STANDALONE_WIN || UNITY_WSA || UNITY_ANDROID) && GLTFAST_PRESENT && KTX_PRESENT
+
+namespace Microsoft.MixedReality.Toolkit.Input
+{
+    /// <summary>
+    /// A helper class which loads and caches controller models fetched from the controller's platform sdk
+    /// </summary>
+    public class ControllerModelLoader
+    {
+        /// <summary>
+        /// A dictionary which caches the controller model gameobject associated with a specified model key
+        /// </summary>
+        public static Dictionary<ulong, GameObject> ControllerModelDictionary = new Dictionary<ulong, GameObject>();
+
+        /// <summary>
+        /// Tries to load the controller model game object from the provided OpenXR ControllerModel.
+        /// Requires the MR OpenXR plugin to work.
+        /// </summary>
+        /// <param name="controllerModelProvider">The OpenXR ControllerModel to loade from</param>
+        /// <returns>A gameobject representing the generated controller model in the scene</returns>
+        public async static Task<GameObject> TryGenerateControllerModelFromPlatformSDK(Handedness handedness)
+        {
+            GameObject gltfGameObject = null;
+
+#if MROPENXR_PRESENT && (UNITY_STANDALONE_WIN || UNITY_WSA || UNITY_ANDROID) && GLTFAST_PRESENT && KTX_PRESENT
+            ControllerModel controllerModelProvider = handedness == Handedness.Left ? ControllerModel.Left :
+                                                        handedness == Handedness.Right ?  ControllerModel.Right :
+                                                        null;
+
+            if (controllerModelProvider == null)
+            {
+                Debug.LogWarning("Controller model provider does not exist for this handedness");
+                return null;
+            }
+
+            if (!controllerModelProvider.TryGetControllerModelKey(out ulong modelKey))
+            {
+                Debug.LogWarning("Failed to obtain controller model key from platform.");
+                return null;
+            }
+
+            if (ControllerModelDictionary.TryGetValue(modelKey, out gltfGameObject))
+            {
+                gltfGameObject.SetActive(true);
+                return gltfGameObject;
+            }
+
+            byte[] modelStream = await controllerModelProvider.TryGetControllerModel(modelKey);
```

Comment:
Not sure if this is 100% in the scope of this PR, but `TryGetControllerModel` seems more like a `GetControllerModelAsync` to me.

Reply:
No strong feelings on the name, but that's also part of the MR OpenXR plugin's API rather than MRTK's

Reply:
I see!